File: odd-even-linked-list.py
from typing import List

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def oddEvenList(self, head: ListNode) -> ListNode:
        if head == None:
            return None
        else:
            odd_head = head
            even_head = head.next
            odd_tail = odd_head
            even_tail = even_head
            while (even_tail != None and even_tail.next != None):
                odd_tail.next = even_tail.next
                odd_tail = odd_tail.next
                even_tail.next = odd_tail.next
                even_tail = even_tail.next
            odd_tail.next = even_head
            return odd_head

File: Odd Even Linked List/odd-even-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
      if (head == nullptr) return nullptr;
      else
	{
	  auto odd_head = head;
	  auto even_head = head->next;
	  auto odd_tail = odd_head;
	  auto even_tail = even_head;
	  while ((even_tail != nullptr) && (even_tail->next != nullptr))
	    {
	      odd_tail.next = even_tail->next;
	      odd_tail = odd_tail->next;
	      even_tail->next = odd_tail->next;
	      even_tail = even_tail->next;
	    }
	  odd_tail->next = even_head;
	  return odd_head;
	}
    }
};
